File: app/src/main/java/lv/luhmirin/sampleapp/login/LoginPresenter.java
package lv.luhmirin.sampleapp.login;

import android.support.annotation.NonNull;

class LoginPresenter {

    private final LoginViewContract contract;

    LoginPresenter(@NonNull LoginViewContract contract) {
        this.contract = contract;
    }

    void login(@NonNull String email, @NonNull String password) {
        contract.cleanErrors();

        if (checkEmail(email) && checkPassword(password)) {
            contract.loginSuccess();
        }
    }


    private boolean checkEmail(@NonNull String email) {
        if (email.isEmpty()) {
            contract.noEmailError();
            return false;
        } else if (notValidEmail(email)) {
            contract.invalidEmailError();
            return false;
        }
        return true;
    }

    private boolean notValidEmail(@NonNull String email) {
        return !email.contains("@") || email.length() < 5;
    }


    private boolean checkPassword(@NonNull String password) {
        if (password.isEmpty()) {
            contract.noPasswordError();
            return false;
        } else if (notValidPassword(password)) {
            contract.invalidPasswordError();
            return false;
        }
        return true;
    }

    private boolean notValidPassword(@NonNull String password) {
        return password.length() < 4;
    }
}

File: app/src/main/kotlin/lv/luhmirin/sampleapp/login/LoginPresenter.kt
package lv.luhmirin.sampleapp.login

internal class LoginPresenter(private val contract: LoginViewContract) {

    fun login(email: String, password: String) {
        contract.cleanErrors()

        if (email.checkEmail() && password.checkPassword()) {
            contract.loginSuccess()
        }
    }

    private fun String.checkEmail(): Boolean {
        if (this.isEmpty()) {
            contract.noEmailError()
            return false
        } else if (this.isNotValidEmail()) {
            contract.invalidEmailError()
            return false
        }
        return true
    }

    private fun String.isNotValidEmail() = !this.contains("@") || this.length < 5


    private fun String.checkPassword(): Boolean {
        if (this.isEmpty()) {
            contract.noPasswordError()
            return false
        } else if (this.isNotValidPassword()) {
            contract.invalidPasswordError()
            return false
        }
        return true
    }

    private fun String.isNotValidPassword() = this.length < 4

}
